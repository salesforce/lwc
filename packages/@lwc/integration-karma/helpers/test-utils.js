/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */

window.TestUtils = (function (lwc, jasmine, beforeAll) {
    function pass() {
        return {
            pass: true,
        };
    }

    function fail(message) {
        return {
            pass: false,
            message: message,
        };
    }

    // TODO [#869]: Replace this custom spy with standard spyOn jasmine spy when logWarning doesn't use console.group
    // anymore. On IE11 console.group has a different behavior when the F12 inspector is attached to the page.
    function spyConsole() {
        var originalConsole = window.console;

        var calls = {
            log: [],
            warn: [],
            error: [],
            group: [],
            groupEnd: [],
        };

        window.console = {
            log: function () {
                calls.log.push(Array.prototype.slice.call(arguments));
            },
            warn: function () {
                calls.warn.push(Array.prototype.slice.call(arguments));
            },
            error: function () {
                calls.error.push(Array.prototype.slice.call(arguments));
            },
            group: function () {
                calls.group.push(Array.prototype.slice.call(arguments));
            },
            groupEnd: function () {
                calls.groupEnd.push(Array.prototype.slice.call(arguments));
            },
        };

        return {
            calls: calls,
            reset: function () {
                window.console = originalConsole;
            },
        };
    }

    function formatConsoleCall(args) {
        return args.map(String).join(' ');
    }

    // TODO [#869]: Improve lookup logWarning doesn't use console.group anymore.
    function consoleDevMatcherFactory(methodName, internalMethodName) {
        return function consoleDevMatcher() {
            return {
                negativeCompare: function negativeCompare(actual) {
                    var spy = spyConsole();
                    try {
                        actual();
                    } finally {
                        spy.reset();
                    }

                    var callsArgs = spy.calls[internalMethodName || methodName];
                    var formattedCalls = callsArgs
                        .map(function (arg) {
                            return '"' + formatConsoleCall(arg) + '"';
                        })
                        .join(', ');

                    if (callsArgs.length === 0) {
                        return {
                            pass: true,
                        };
                    }
                    return {
                        pass: false,
                        message: function () {
                            return 'Expect no message but received:\n' + formattedCalls;
                        },
                    };
                },
                compare: function compare(actual, expectedMessages) {
                    function matchMessage(message, expectedMessage) {
                        if (typeof expectedMessage === 'string') {
                            return message === expectedMessage;
                        } else {
                            return expectedMessage.test(message);
                        }
                    }

                    if (!Array.isArray(expectedMessages)) {
                        expectedMessages = [expectedMessages];
                    }

                    if (typeof actual !== 'function') {
                        throw new Error('Expected function to throw error.');
                    } else if (
                        expectedMessages.some(function (message) {
                            return typeof message !== 'string' && !(message instanceof RegExp);
                        })
                    ) {
                        throw new Error(
                            'Expected a string or a RegExp to compare the thrown error against, or an array of such.'
                        );
                    }

                    var spy = spyConsole();

                    try {
                        actual();
                    } finally {
                        spy.reset();
                    }

                    var callsArgs = spy.calls[internalMethodName || methodName];
                    var formattedCalls = callsArgs
                        .map(function (callArgs) {
                            return '"' + formatConsoleCall(callArgs) + '"';
                        })
                        .join(', ');

                    if (process.env.NODE_ENV === 'production') {
                        if (callsArgs.length !== 0) {
                            fail(
                                'Expected console.' +
                                    methodName +
                                    ' to never called in production mode, but it was called ' +
                                    callsArgs.length +
                                    ' with ' +
                                    formattedCalls +
                                    '.'
                            );
                        } else {
                            return pass();
                        }
                    } else {
                        if (callsArgs.length === 0) {
                            return fail(
                                'Expected console.' +
                                    methodName +
                                    ' to called with ' +
                                    JSON.stringify(expectedMessages) +
                                    ', but was never called.'
                            );
                        } else {
                            if (callsArgs.length !== expectedMessages.length) {
                                return fail(
                                    'Expected console.' +
                                        methodName +
                                        ' to be called ' +
                                        expectedMessages.length +
                                        ' time(s), but was called ' +
                                        callsArgs.length +
                                        ' time(s).'
                                );
                            }
                            for (var i = 0; i < callsArgs.length; i++) {
                                var callsArg = callsArgs[i];
                                var expectedMessage = expectedMessages[i];
                                var actualMessage = formatConsoleCall(callsArg);
                                if (!matchMessage(actualMessage, expectedMessage)) {
                                    return fail(
                                        'Expected console.' +
                                            methodName +
                                            ' to be called with "' +
                                            expectedMessage +
                                            '", but was called with "' +
                                            actualMessage +
                                            '".'
                                    );
                                }
                            }
                            return pass();
                        }
                    }
                },
            };
        };
    }

    function errorMatcherFactory(errorListener) {
        return function toThrowError() {
            return {
                compare: function (actual, expectedErrorCtor, expectedMessage) {
                    function matchMessage(message) {
                        if (typeof expectedMessage === 'undefined') {
                            return true;
                        } else if (typeof expectedMessage === 'string') {
                            return message === expectedMessage;
                        } else {
                            return expectedMessage.test(message);
                        }
                    }

                    function matchError(error) {
                        return error instanceof expectedErrorCtor && matchMessage(error.message);
                    }

                    function throwDescription(thrown) {
                        return thrown.name + ' with message "' + thrown.message + '"';
                    }

                    if (typeof expectedMessage === 'undefined') {
                        if (typeof expectedErrorCtor === 'undefined') {
                            // 0 arguments provided
                            expectedMessage = undefined;
                            expectedErrorCtor = Error;
                        } else {
                            // 1 argument provided
                            expectedMessage = expectedErrorCtor;
                            expectedErrorCtor = Error;
                        }
                    }

                    if (typeof actual !== 'function') {
                        throw new Error('Expected function to throw error.');
                    } else if (
                        expectedErrorCtor !== Error &&
                        !(expectedErrorCtor.prototype instanceof Error)
                    ) {
                        throw new Error('Expected an error constructor.');
                    } else if (
                        typeof expectedMessage !== 'undefined' &&
                        typeof expectedMessage !== 'string' &&
                        !(expectedMessage instanceof RegExp)
                    ) {
                        throw new Error(
                            'Expected a string or a RegExp to compare the thrown error against.'
                        );
                    }

                    var thrown = errorListener(actual);

                    if (process.env.NODE_ENV === 'production') {
                        if (thrown !== undefined) {
                            return fail(
                                'Expected function not to throw an error in production mode, but it threw ' +
                                    throwDescription(thrown) +
                                    '.'
                            );
                        } else {
                            return pass();
                        }
                    } else {
                        if (thrown === undefined) {
                            return fail(
                                'Expected function to throw an ' +
                                    expectedErrorCtor.name +
                                    ' error in development mode"' +
                                    (expectedMessage ? 'with message ' + expectedMessage : '') +
                                    '".'
                            );
                        } else if (!matchError(thrown)) {
                            return fail(
                                'Expected function to throw an ' +
                                    expectedErrorCtor.name +
                                    ' error in development mode "' +
                                    (expectedMessage ? 'with message ' + expectedMessage : '') +
                                    '", but it threw ' +
                                    throwDescription(thrown) +
                                    '.'
                            );
                        } else {
                            return pass();
                        }
                    }
                },
            };
        };
    }

    // Listen for errors thrown directly by the callback
    function directErrorListener(callback) {
        try {
            callback();
        } catch (error) {
            return error;
        }
    }

    // Listen for errors using window.addEventListener('error')
    function windowErrorListener(callback) {
        var error;
        function onError(event) {
            event.preventDefault(); // don't log the error
            error = event.error;
        }

        // Prevent jasmine from handling the global error. There doesn't seem to be another
        // way to disable this behavior: https://github.com/jasmine/jasmine/pull/1860
        var originalOnError = window.onerror;
        window.onerror = null;
        window.addEventListener('error', onError);

        try {
            callback();
        } finally {
            window.onerror = originalOnError;
            window.removeEventListener('error', onError);
        }
        return error;
    }

    // For errors we expect to be thrown in the connectedCallback() phase
    // of a custom element, there are two possibilities:
    // 1) We're using non-native lifecycle callbacks, so the error is thrown synchronously
    // 2) We're using native lifecycle callbacks, so the error is thrown asynchronously and can
    //    only be caught with window.addEventListener('error')
    function customElementConnectedErrorListener(callback) {
        return window.lwcRuntimeFlags.ENABLE_NATIVE_CUSTOM_ELEMENT_LIFECYCLE
            ? windowErrorListener(callback)
            : directErrorListener(callback);
    }

    var customMatchers = {
        toLogErrorDev: consoleDevMatcherFactory('error'),
        toLogWarningDev: consoleDevMatcherFactory('warn'),
        toThrowErrorDev: errorMatcherFactory(directErrorListener),
        toThrowConnectedError: errorMatcherFactory(customElementConnectedErrorListener),
    };

    beforeAll(function () {
        jasmine.addMatchers(customMatchers);
    });

    function extractDataIds(root) {
        var nodes = {};

        function processElement(elm) {
            if (elm.hasAttribute('data-id')) {
                nodes[elm.getAttribute('data-id')] = elm;
            }

            if (elm.shadowRoot) {
                Object.assign(nodes, extractShadowDataIds(elm.shadowRoot));
            }
        }

        function acceptNode() {
            return NodeFilter.FILTER_ACCEPT;
        }

        // Work around Internet Explorer wanting a function instead of an object. IE also *requires* this argument where
        // other browsers don't.
        var safeFilter = acceptNode;
        safeFilter.acceptNode = acceptNode;

        var walker = document.createTreeWalker(root, NodeFilter.SHOW_ELEMENT, safeFilter, false);

        processElement(root);

        var elm;
        while ((elm = walker.nextNode())) {
            processElement(elm);
        }

        return nodes;
    }

    function extractShadowDataIds(shadowRoot) {
        var nodes = {};

        // Add the shadow root here even if they don't have [data-id] attributes. This reference is
        // subsequently used to add event listeners.
        var dataId = shadowRoot.host.getAttribute('data-id');
        if (dataId) {
            nodes[dataId + '.shadowRoot'] = shadowRoot;
        }

        // We can't use a TreeWalker directly on the ShadowRoot since with synthetic shadow the ShadowRoot is not an
        // actual DOM nodes. So we need to iterate over the children manually and run the tree walker on each child.
        for (var i = 0; i < shadowRoot.childNodes.length; i++) {
            var child = shadowRoot.childNodes[i];
            Object.assign(nodes, extractDataIds(child));
        }

        return nodes;
    }

    var register = {};
    function load(id) {
        return Promise.resolve(register[id]);
    }

    function registerForLoad(name, Ctor) {
        register[name] = Ctor;
    }
    function clearRegister() {
        register = {};
    }

    // #986 - childNodes on the host element returns a fake shadow comment node on IE11 for debugging purposed. This method
    // filters this node.
    function getHostChildNodes(host) {
        return Array.prototype.slice.call(host.childNodes).filter(function (n) {
            return !(n.nodeType === Node.COMMENT_NODE && n.tagName.startsWith('#shadow-root'));
        });
    }

    function isSyntheticShadowRootInstance(sr) {
        return Boolean(sr && sr.synthetic);
    }

    function isNativeShadowRootInstance(sr) {
        return Boolean(sr && !sr.synthetic);
    }

    // Providing overridable hooks for tests
    var sanitizeHtmlContentHook = function () {
        throw new Error('sanitizeHtmlContent hook must be implemented.');
    };

    LWC.setHooks({
        sanitizeHtmlContent: function (content) {
            return sanitizeHtmlContentHook(content);
        },
    });

    function getHooks() {
        return {
            sanitizeHtmlContent: sanitizeHtmlContentHook,
        };
    }

    function setHooks(hooks) {
        if (hooks.sanitizeHtmlContent) {
            sanitizeHtmlContentHook = hooks.sanitizeHtmlContent;
        }
    }

    // This mapping should be kept up-to-date with the mapping in @lwc/shared -> aria.ts
    var ariaPropertiesMapping = {
        ariaAutoComplete: 'aria-autocomplete',
        ariaChecked: 'aria-checked',
        ariaCurrent: 'aria-current',
        ariaDisabled: 'aria-disabled',
        ariaExpanded: 'aria-expanded',
        ariaHasPopup: 'aria-haspopup',
        ariaHidden: 'aria-hidden',
        ariaInvalid: 'aria-invalid',
        ariaLabel: 'aria-label',
        ariaLevel: 'aria-level',
        ariaMultiLine: 'aria-multiline',
        ariaMultiSelectable: 'aria-multiselectable',
        ariaOrientation: 'aria-orientation',
        ariaPressed: 'aria-pressed',
        ariaReadOnly: 'aria-readonly',
        ariaRequired: 'aria-required',
        ariaSelected: 'aria-selected',
        ariaSort: 'aria-sort',
        ariaValueMax: 'aria-valuemax',
        ariaValueMin: 'aria-valuemin',
        ariaValueNow: 'aria-valuenow',
        ariaValueText: 'aria-valuetext',
        ariaLive: 'aria-live',
        ariaRelevant: 'aria-relevant',
        ariaAtomic: 'aria-atomic',
        ariaBusy: 'aria-busy',
        ariaActiveDescendant: 'aria-activedescendant',
        ariaControls: 'aria-controls',
        ariaDescribedBy: 'aria-describedby',
        ariaFlowTo: 'aria-flowto',
        ariaLabelledBy: 'aria-labelledby',
        ariaOwns: 'aria-owns',
        ariaPosInSet: 'aria-posinset',
        ariaSetSize: 'aria-setsize',
        ariaColCount: 'aria-colcount',
        ariaColSpan: 'aria-colspan',
        ariaColIndex: 'aria-colindex',
        ariaDetails: 'aria-details',
        ariaErrorMessage: 'aria-errormessage',
        ariaKeyShortcuts: 'aria-keyshortcuts',
        ariaModal: 'aria-modal',
        ariaPlaceholder: 'aria-placeholder',
        ariaRoleDescription: 'aria-roledescription',
        ariaRowCount: 'aria-rowcount',
        ariaRowIndex: 'aria-rowindex',
        ariaRowSpan: 'aria-rowspan',
        role: 'role',
    };

    var ariaProperties = Object.keys(ariaPropertiesMapping);

    // Can't use Object.values because we need to support IE11
    var ariaAttributes = [];
    for (let i = 0; i < ariaProperties.length; i++) {
        ariaAttributes.push(ariaPropertiesMapping[ariaProperties[i]]);
    }

    return {
        clearRegister: clearRegister,
        extractDataIds: extractDataIds,
        extractShadowDataIds: extractShadowDataIds,
        getHostChildNodes: getHostChildNodes,
        isNativeShadowRootInstance: isNativeShadowRootInstance,
        isSyntheticShadowRootInstance: isSyntheticShadowRootInstance,
        load: load,
        registerForLoad: registerForLoad,
        getHooks: getHooks,
        setHooks: setHooks,
        spyConsole: spyConsole,
        customElementConnectedErrorListener: customElementConnectedErrorListener,
        ariaPropertiesMapping: ariaPropertiesMapping,
        ariaProperties: ariaProperties,
        ariaAttributes: ariaAttributes,
    };
})(LWC, jasmine, beforeAll);
