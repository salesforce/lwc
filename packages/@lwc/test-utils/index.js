/**
 * Returns the shadowRoot property of a given Lightning web component.
 *
 * Use this utility instead of directly accessing the the element's ShadowRoot
 * to future-proof your test logic as LWC's Shadow DOM API implementation
 * evolves over time.
 *
 * @param {LWCElement} element The Lightning web component element to retrieve
 * the shadowRoot property off of
 * @returns {ShadowRoot} The shadow root of the given element
 */
module.exports.getShadowRoot = function(element) {
    if (!element || !element.$$ShadowRoot$$) {
        const tagName = element && element.tagName && element.tagName.toLowerCase();
        throw new Error(`Attempting to retrieve the shadow root of '${tagName || element}' but no shadowRoot property found`);
    }
    return element.$$ShadowRoot$$;
}
