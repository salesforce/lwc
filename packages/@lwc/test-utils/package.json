{
  "name": "@lwc/test-utils",
  "version": "0.33.9",
  "description": "Test utilities for LWC",
  "main": "index.js",
  "license": "MIT",
  "keywords": [
    "lwc"
  ]
}
