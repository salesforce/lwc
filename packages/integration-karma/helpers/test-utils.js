/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */

window.TestUtils = (function (lwc, jasmine, beforeAll) {
    function pass() {
        return {
            pass: true,
        };
    }

    function fail(message) {
        return {
            pass: false,
            message: message,
        };
    }

    // TODO [#869]: Replace this custom spy with standard spyOn jasmine spy when logWarning doesn't use console.group
    // anymore. On IE11 console.group has a different behavior when the F12 inspector is attached to the page.
    function spyConsole() {
        var originalConsole = window.console;

        var calls = {
            log: [],
            warn: [],
            error: [],
            group: [],
            groupEnd: [],
        };

        window.console = {
            log: function () {
                calls.log.push(Array.prototype.slice.call(arguments));
            },
            warn: function () {
                calls.warn.push(Array.prototype.slice.call(arguments));
            },
            error: function () {
                calls.error.push(Array.prototype.slice.call(arguments));
            },
            group: function () {
                calls.group.push(Array.prototype.slice.call(arguments));
            },
            groupEnd: function () {
                calls.groupEnd.push(Array.prototype.slice.call(arguments));
            },
        };

        return {
            calls: calls,
            reset: function () {
                window.console = originalConsole;
            },
        };
    }

    function formatConsoleCall(args) {
        return args.map(String).join(' ');
    }

    // TODO [#869]: Improve lookup logWarning doesn't use console.group anymore.
    function consoleDevMatcherFactory(methodName, internalMethodName) {
        return function consoleDevMatcher() {
            return {
                negativeCompare: function negativeCompare(actual) {
                    var spy = spyConsole();
                    try {
                        actual();
                    } finally {
                        spy.reset();
                    }

                    var callsArgs = spy.calls[internalMethodName || methodName];
                    var formattedCalls = callsArgs
                        .map(function (arg) {
                            return '"' + formatConsoleCall(arg) + '"';
                        })
                        .join(', ');

                    if (callsArgs.length === 0) {
                        return {
                            pass: true,
                        };
                    }
                    return {
                        pass: false,
                        message: function () {
                            return 'Expect no message but received:\n' + formattedCalls;
                        },
                    };
                },
                compare: function compare(actual, expectedMessage) {
                    function matchMessage(message) {
                        if (typeof expectedMessage === 'string') {
                            return message === expectedMessage;
                        } else {
                            return expectedMessage.test(message);
                        }
                    }

                    if (typeof actual !== 'function') {
                        throw new Error('Expected function to throw error.');
                    } else if (
                        typeof expectedMessage !== 'string' &&
                        !(expectedMessage instanceof RegExp)
                    ) {
                        throw new Error(
                            'Expected a string or a RegExp to compare the thrown error against.'
                        );
                    }

                    var spy = spyConsole();

                    try {
                        actual();
                    } finally {
                        spy.reset();
                    }

                    var callsArgs = spy.calls[internalMethodName || methodName];
                    var formattedCalls = callsArgs
                        .map(function (callArgs) {
                            return '"' + formatConsoleCall(callArgs) + '"';
                        })
                        .join(', ');

                    if (process.env.NODE_ENV === 'production') {
                        if (callsArgs.length !== 0) {
                            fail(
                                'Expected console.' +
                                    methodName +
                                    ' to never called in production mode, but it was called ' +
                                    callsArgs.length +
                                    ' with ' +
                                    formattedCalls +
                                    '.'
                            );
                        } else {
                            return pass();
                        }
                    } else {
                        if (callsArgs.length === 0) {
                            return fail(
                                'Expected console.' +
                                    methodName +
                                    ' to called once with "' +
                                    expectedMessage +
                                    '", but was never called.'
                            );
                        } else if (callsArgs.length === 1) {
                            var actualMessage = formatConsoleCall(callsArgs[0]);

                            if (!matchMessage(actualMessage)) {
                                return fail(
                                    'Expected console.' +
                                        methodName +
                                        ' to be called with "' +
                                        expectedMessage +
                                        '", but was called with "' +
                                        actualMessage +
                                        '".'
                                );
                            } else {
                                return pass();
                            }
                        } else {
                            return fail(
                                'Expected console.' +
                                    methodName +
                                    ' to never called, but it was called ' +
                                    callsArgs.length +
                                    ' with ' +
                                    formattedCalls +
                                    '.'
                            );
                        }
                    }
                },
            };
        };
    }

    var customMatchers = {
        toLogErrorDev: consoleDevMatcherFactory('error'),
        toThrowErrorDev: function toThrowErrorDev() {
            return {
                compare: function (actual, expectedErrorCtor, expectedMessage) {
                    function matchMessage(message) {
                        if (typeof expectedMessage === 'string') {
                            return message === expectedMessage;
                        } else {
                            return expectedMessage.test(message);
                        }
                    }

                    function matchError(error) {
                        return error instanceof expectedErrorCtor && matchMessage(error.message);
                    }

                    function throwDescription(thrown) {
                        return thrown.name + ' with message "' + thrown.message + '"';
                    }

                    if (typeof actual !== 'function') {
                        throw new Error('Expected function to throw error.');
                    } else if (
                        typeof actual !== 'function' ||
                        expectedErrorCtor.prototype instanceof Error
                    ) {
                        throw new Error('Expected an error constructor.');
                    } else if (
                        typeof expectedMessage !== 'string' &&
                        !(expectedMessage instanceof RegExp)
                    ) {
                        throw new Error(
                            'Expected a string or a RegExp to compare the thrown error against.'
                        );
                    }

                    let thrown;

                    try {
                        actual();
                    } catch (error) {
                        thrown = error;
                    }

                    if (process.env.NODE_ENV === 'production') {
                        if (thrown !== undefined) {
                            return fail(
                                'Expected function not to throw an error in production mode, but it threw ' +
                                    throwDescription(thrown) +
                                    '.'
                            );
                        } else {
                            return pass();
                        }
                    } else {
                        if (thrown === undefined) {
                            return fail(
                                'Expected function to throw an ' +
                                    expectedErrorCtor.name +
                                    ' error in development mode with message "' +
                                    expectedMessage +
                                    '".'
                            );
                        } else if (!matchError(thrown)) {
                            return fail(
                                'Expected function to throw an ' +
                                    expectedErrorCtor.name +
                                    ' error in development mode with message "' +
                                    expectedMessage +
                                    '", but it threw ' +
                                    throwDescription(thrown) +
                                    '.'
                            );
                        } else {
                            return pass();
                        }
                    }
                },
            };
        },
    };

    beforeAll(function () {
        jasmine.addMatchers(customMatchers);
    });

    function extractDataIds(root) {
        var nodes = {};

        function processElement(elm) {
            if (elm.hasAttribute('data-id')) {
                nodes[elm.getAttribute('data-id')] = elm;
            }

            if (elm.shadowRoot) {
                Object.assign(nodes, extractShadowDataIds(elm.shadowRoot));
            }
        }

        function acceptNode() {
            return NodeFilter.FILTER_ACCEPT;
        }

        // Work around Internet Explorer wanting a function instead of an object. IE also *requires* this argument where
        // other browsers don't.
        var safeFilter = acceptNode;
        safeFilter.acceptNode = acceptNode;

        var walker = document.createTreeWalker(root, NodeFilter.SHOW_ELEMENT, safeFilter, false);

        processElement(root);

        var elm;
        while ((elm = walker.nextNode())) {
            processElement(elm);
        }

        return nodes;
    }

    function extractShadowDataIds(shadowRoot) {
        var nodes = {};

        // Add the shadow root here even if they don't have [data-id] attributes. This reference is
        // subsequently used to add event listeners.
        var dataId = shadowRoot.host.getAttribute('data-id');
        if (dataId) {
            nodes[dataId + '.shadowRoot'] = shadowRoot;
        }

        // We can't use a TreeWalker directly on the ShadowRoot since with synthetic shadow the ShadowRoot is not an
        // actual DOM nodes. So we need to iterate over the children manually and run the tree walker on each child.
        for (var i = 0; i < shadowRoot.childNodes.length; i++) {
            var child = shadowRoot.childNodes[i];
            Object.assign(nodes, extractDataIds(child));
        }

        return nodes;
    }

    var register = {};
    function load(id) {
        return Promise.resolve(register[id]);
    }

    function registerForLoad(name, Ctor) {
        register[name] = Ctor;
    }
    function clearRegister() {
        register = {};
    }

    // #986 - childNodes on the host element returns a fake shadow comment node on IE11 for debugging purposed. This method
    // filters this node.
    function getHostChildNodes(host) {
        return Array.prototype.slice.call(host.childNodes).filter(function (n) {
            return !(n.nodeType === Node.COMMENT_NODE && n.tagName.startsWith('#shadow-root'));
        });
    }
    return {
        registerForLoad: registerForLoad,
        clearRegister: clearRegister,
        load: load,
        extractDataIds: extractDataIds,
        extractShadowDataIds: extractShadowDataIds,
        getHostChildNodes: getHostChildNodes,
    };
})(LWC, jasmine, beforeAll);
