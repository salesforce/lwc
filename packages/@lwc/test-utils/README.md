# lwc-test-utils

A collection of utility functions to assist in unit testing Lightning web components.

## Usage

Usage of this library assumes you have the a Lightning Web Component development and test environment setup. To get started with testing your Lightning web components see the [documentation website](https://internal.lwcjs.org/guide/testing.html).

### Direct Import

Add this package as a devDependency and import the utils directly into your test.

```bash
yarn add -D '@lwc/test-utils'
```

In your test:

```js
import { getShadowRoot } from '@lwc/test-utils';

// create element under test and query for custom element in the Shadow DOM
const element = createElement('my-custom-component', { is: MyCustomComponent });
document.body.appendChild(element);
const shadowRoot = getShadowRoot(element);
const fancyButton = shadowRoot.querySelector('fancy-button');
```

### Jest Preset Configuration

This util also comes pre-packed in `lwc-jest-preset`.

```bash
yarn add -D 'lwc-jest-preset'
```

In your `package.json`:

```json
"jest": {
    "preset": "lwc-jest-preset",
    "moduleNameMapper": {
        "^(mynamespace)-(.+)$": "<rootDir>/src/$1/$2/$2"
    }
}
```

Or in `jest.config.js`:

```js
const PRESET_CONFIG = require('lwc-jest-preset');
module.exports = {
    ...PRESET_CONFIG,
    moduleNameMapper: {
        ...PRESET_CONFIG.moduleNameMapper,
        '^(mynamespace)-(.+)$': '<rootDir>/src/$1/$2/$2'
    },
};
```
