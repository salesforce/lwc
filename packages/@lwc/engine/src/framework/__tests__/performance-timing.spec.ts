/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */
import { compileTemplate } from 'test-utils';

declare var global: Global;

interface Global {
    performance: UserTiming;
}

interface UserTiming {
    mark(name: string): void;
    measure(label: string, name: string): void;
    clearMarks(name: string): void;
    clearMeasures(label: string): void;
}

interface Mark {
    name: string;
    label: null | string;
    children: Mark[];
    parent: Mark;
}

class FlameChart {
    activeMark: Mark;

    constructor() {
        this.reset();
    }

    reset() {
        this.activeMark = {
            name: '<ROOT>',
            label: '<ROOT>',
            children: [],
            parent: null,
        } as any;
    }

    injectPolyfill() {
        Object.assign(global.performance, {
            mark: (name: string) => {
                const mark: Mark = {
                    name,
                    label: null,
                    children: [],
                    parent: this.activeMark,
                };

                if (this.activeMark) {
                    this.activeMark.children.push(mark);
                }

                this.activeMark = mark;
            },

            measure: (label: string, mark: string) => {
                if (!this.activeMark) {
                    throw new Error(`Unexpected measure ${label}, no matching mark for ${mark}`);
                } else if (this.activeMark.name !== mark) {
                    throw new Error(`Unexpected measure ${label}, expected ${this.activeMark.name} received ${mark}`);
                }

                this.activeMark.label = label;
                this.activeMark = this.activeMark.parent;
            },

            clearMarks: () => {},

            clearMeasures: () => {},
        });
    }

    buildFlamechart(
        mark: Mark = this.activeMark,
        indent: number = 0
    ): string {
        return [
            '\t'.repeat(indent) + mark.label,
            ...mark.children.map(c => this.buildFlamechart(c, indent + 1)),
        ].join('\n');
    }

    toString() {
        return this.buildFlamechart();
    }
}

// TODO: Those test have are not exercising the expected logic.
// https://github.com/salesforce/lwc/commit/9cd09d41bc50907b2ee6f9ada3467a0de237db81#diff-b6e048115546dca1a7c9a1585e9b7a0c
describe.skip('performance-timing', () => {
    let createElement: any;
    let registerTemplate: any;
    let LightningElement: any;
    let flamechart: FlameChart;

    beforeEach(() => {
        // Make sure to reset module cache between each test to ensure to reset the uid.
        jest.resetModules();

        flamechart = new FlameChart();
        flamechart.injectPolyfill();

        const lwc = require('../main.ts');
        createElement = lwc.createElement;
        LightningElement = lwc.LightningElement;
        registerTemplate = lwc.registerTemplate;
    });

    it('captures component constructor', () => {
        class Foo extends LightningElement {}

        const elm = createElement('x-foo', { is: Foo });
        document.body.appendChild(elm);

        expect(flamechart.toString()).toMatchSnapshot();
    });

    it('captures all lifecycle hooks', () => {
        class Foo extends LightningElement {
            connectedCallback() {}
            renderedCallback() {}
            disconnectedCallback() {}
        }

        const elm = createElement('x-foo', { is: Foo });
        document.body.appendChild(elm);
        document.body.removeChild(elm);

        expect(flamechart.toString()).toMatchSnapshot();
    });

    it('captures nested component tree', () => {
        const barTmpl = compileTemplate(`<template>{foo}<span></span></template>`);
        class Bar extends LightningElement {
            render() {
                return registerTemplate(barTmpl);
            }
        }

        const fooTmpl = compileTemplate(`
            <template>
                <x-bar></x-bar>
                <x-bar></x-bar>
            </template>
        `, {
            modules: {
                'x-bar': Bar,
            }
        });
        class Foo extends LightningElement {
            render() {
                return registerTemplate(fooTmpl);
            }
        }

        const elm = createElement('x-foo', { is: Foo });
        document.body.appendChild(elm);

        expect(flamechart.toString()).toMatchSnapshot();
    });

    it('recovers from errors', () => {
        class Foo extends LightningElement {
            render() {
                throw new Error('Nooo!');
            }
        }

        const elm = createElement('x-foo', { is: Foo });

        try {
            document.body.appendChild(elm);
        } catch (err) {
            expect(err.message).toBe('Nooo!');
        }

        expect(flamechart.toString()).toMatchSnapshot();
    });

    it('captures error callback', () => {
        class Bar extends LightningElement {
            render() {
                throw new Error('Noo!');
            }
        }

        const fooTmpl = compileTemplate(`
            <template>
                <x-bar></x-bar>
            </template>
        `, {
            modules: {
                'x-bar': Bar,
            }
        });
        class Foo extends LightningElement {
            render() {
                return registerTemplate(fooTmpl);
            }
            errorCallback() {}
        }

        const elm = createElement('x-foo', { is: Foo });
        document.body.appendChild(elm);

        expect(flamechart.toString()).toMatchSnapshot();
    });

    it('skips parent measurement when children component props change', () => {
        let bar: Bar;
        let baz: Baz;

        const barTmpl = compileTemplate(`<template>{state}</template>`);
        class Bar extends LightningElement {
            state: boolean = false;

            connectedCallback() {
                bar = this;
            }

            render() {
                return registerTemplate(barTmpl);
            }

            static track = {
                state: {}
            };
        }

        const bazTmpl = compileTemplate(`<template>{state}</template>`);
        class Baz extends LightningElement {
            state: boolean = false;

            connectedCallback() {
                baz = this;
            }

            render() {
                return registerTemplate(bazTmpl);
            }

            static track = {
                state: {}
            };
        }

        const fooTmpl = compileTemplate(`
            <template>
                <x-bar></x-bar>
                <x-baz></x-baz>
            </template>
        `, {
            modules: {
                'x-bar': Bar,
                'x-baz': Baz,
            }
        });
        class Foo extends LightningElement {
            render() {
                return registerTemplate(fooTmpl);
            }
        }

        const elm = createElement('x-foo', { is: Foo });
        document.body.appendChild(elm);

        flamechart.reset();

        bar.state = true;
        baz.state = true;

        return Promise.resolve().then(() => (
            expect(flamechart.toString()).toMatchSnapshot()
        ));
    });
});
