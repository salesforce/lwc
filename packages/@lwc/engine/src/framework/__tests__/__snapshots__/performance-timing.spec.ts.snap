// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`performance-timing captures all lifecycle hooks 1`] = `"<ROOT>"`;

exports[`performance-timing captures component constructor 1`] = `"<ROOT>"`;

exports[`performance-timing captures error callback 1`] = `"<ROOT>"`;

exports[`performance-timing captures nested component tree 1`] = `"<ROOT>"`;

exports[`performance-timing recovers from errors 1`] = `"<ROOT>"`;

exports[`performance-timing skips parent measurement when children component props change 1`] = `"<ROOT>"`;
